cmd/up/exec.go

```diff
@@ -16,27 +16,268 @@ package up
 import (
 	"bytes"
 	"context"
+	"encoding/base64"
+	"encoding/json"
 	"fmt"
+	"io"
 	"os"
+	"os/exec"
+	"path/filepath"
 	"strings"
 
+	"github.com/creack/pty"
 	"github.com/okteto/okteto/cmd/utils"
+	"github.com/okteto/okteto/pkg/cmd/pipeline"
 	"github.com/okteto/okteto/pkg/config"
+	"github.com/okteto/okteto/pkg/constants"
 	oktetoErrors "github.com/okteto/okteto/pkg/errors"
 	"github.com/okteto/okteto/pkg/k8s/apps"
-	"github.com/okteto/okteto/pkg/k8s/exec"
+	"github.com/okteto/okteto/pkg/k8s/configmaps"
+	k8sExec "github.com/okteto/okteto/pkg/k8s/exec"
 	"github.com/okteto/okteto/pkg/k8s/pods"
 	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+	"github.com/okteto/okteto/pkg/okteto"
+	"github.com/okteto/okteto/pkg/registry"
 	"github.com/okteto/okteto/pkg/ssh"
+	"github.com/okteto/okteto/pkg/types"
+	"k8s.io/client-go/kubernetes"
 )
 
+type devExecutor interface {
+	runCommand(ctx context.Context, cmd []string) error
+}
+
+type hybridExecutor struct {
+	workdir string
+	envs    []string
+}
+
+func (he *hybridExecutor) runCommand(_ context.Context, cmd []string) error {
+	c := exec.Command("bash", "-c", strings.Join(cmd, " "))
+	c.Dir = he.workdir
+	c.Env = append(os.Environ(), he.envs...)
+
+	f, err := pty.Start(c)
+	defer func() {
+		_ = f.Close()
+	}()
+	if err != nil {
+		return err
+	}
+
+	io.Copy(os.Stdout, f)
+	return c.Wait()
+}
+
+type syncExecutor struct {
+	iface      string
+	remotePort int
+}
+
+func (se *syncExecutor) runCommand(ctx context.Context, cmd []string) error {
+	return ssh.Exec(ctx, se.iface, se.remotePort, true, os.Stdin, os.Stdout, os.Stderr, cmd)
+}
+
+func newHybridExecutor(ctx context.Context, up *upContext) (*hybridExecutor, error) {
+	wd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+
+	envsGetter, err := newEnvsGetter(up)
+	if err != nil {
+		return nil, err
+	}
+
+	envs, err := envsGetter.getEnvs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &hybridExecutor{
+		workdir: filepath.Join(wd, up.Dev.Workdir),
```

Comment:
Q: Are we checking that the workdir is a relative path? I didn't see it neither in the serializer nor the `dev.go` file. What if the user set it as an absolute path? I think this scenario could be broken

Reply:
done 38016f667ab2144ab79b7fe6e4faf031d85e687f